src/projection.jl

```diff
@@ -52,93 +128,197 @@ function ProjectTo(x::T) where {T<:Union{<:Tuple,NamedTuple}}
-    return T(map(proj -> proj(dx), project.elements))
-end
-(project::ProjectTo{<:Array})(dx::AbstractArray) = project(collect(dx))
+# ProjectTo(xs::T) where {T<:Array} = ProjectTo{T}(; elements=map(ProjectTo, xs))
+# function (project::ProjectTo{T})(dx::Array) where {T<:Array}
+#     _call(f, x) = f(x)
+#     return T(map(_call, project.elements, dx))
+# end
+# function (project::ProjectTo{T})(dx::AbstractZero) where {T<:Array}
+#     return T(map(proj -> proj(dx), project.elements))
+# end
+# (project::ProjectTo{<:Array})(dx::AbstractArray) = project(collect(dx))
 
 # Arrays{<:Number}: optimized case so we don't need a projector per element
-function ProjectTo(x::T) where {E<:Number,T<:Array{E}}
-    return ProjectTo{T}(; element=ProjectTo(zero(E)), size=size(x))
+function ProjectTo(x::AbstractArray{T,N}) where {T<:Number,N}
+    T == Bool && return ProjectTo(false)
+    sub = ProjectTo(zero(T))
+    return ProjectTo{AbstractArray{eltype(sub),N}}(; element=sub, axes=axes(x))
 end
-(project::ProjectTo{<:Array{T}})(dx::Array) where {T<:Number} = project.element.(dx)
-function (project::ProjectTo{<:Array{T}})(dx::AbstractZero) where {T<:Number}
-    return zeros(T, project.size)
+function (project::ProjectTo{AbstractArray{T,N}})(dx::AbstractArray{S,M}) where {T,S,N,M}
+    dy = S <: T ? dx : broadcast(project.element, dx)
+    if axes(dy) == project.axes
+        return dy
+    else
+        for d in 1:max(N,M)
+            size(dy, d) == length(get(project.axes, d, 1)) || throw(DimensionMismatch("wrong shape!"))
+        end
+        return reshape(dy, project.axes)
+    end
 end
-function (project::ProjectTo{<:Array{T}})(dx::Tangent{<:SubArray}) where {T<:Number}
-    return project(dx.parent)
+# function (project::ProjectTo{<:Array{T}})(dx::Tangent{<:SubArray}) where {T<:Number}
+#     return project(dx.parent)
+# end
+
+# Row vectors -- need a bit more optimising!
+function ProjectTo(x::LinearAlgebra.AdjointAbsVec{T}) where {T<:Number}
+    sub = ProjectTo(parent(x))
+    ProjectTo{Adjoint{eltype(eltype(sub)), eltype(sub)}}(; parent=sub)
 end
+(project::ProjectTo{<:Adjoint})(dx::Adjoint) = adjoint(project.parent(parent(dx)))
+(project::ProjectTo{<:Adjoint})(dx::Transpose) = adjoint(conj(project.parent(parent(dx)))) # might copy twice?
+(project::ProjectTo{<:Adjoint})(dx::AbstractArray) = adjoint(conj(project.parent(vec(dx)))) # not happy!
+
+ProjectTo(x::LinearAlgebra.TransposeAbsVec{T}) where {T<:Number} = error("not yet defined")
 
 # Diagonal
-ProjectTo(x::T) where {T<:Diagonal} = ProjectTo{T}(; diag=ProjectTo(diag(x)))
-(project::ProjectTo{T})(dx::AbstractMatrix) where {T<:Diagonal} = T(project.diag(diag(dx)))
-(project::ProjectTo{T})(dx::AbstractZero) where {T<:Diagonal} = T(project.diag(dx))
+function ProjectTo(x::Diagonal)
+    eltype(x) == Bool && return ProjectTo(false)
+    sub = ProjectTo(diag(x))
+    ProjectTo{Diagonal{eltype(eltype(sub)), eltype(sub)}}(; diag=sub)
```

Comment:
This `{eltype(eltype(sub)), eltype(sub)}}` means you will get `p::ProjectTo{Diagonal{Float64, ...}}`. This is mostly a visual reminder of what eltypes are allowed, since the type there is completely ignored when applying `p(dx)`, the eltype which matters is encoded in `p.diag.element`.

So possibly this is just clutter, and making it `p::ProjectTo{Diagonal, ...}` would be cleaner. That would match the `T.name.wrapper` behaviour for other structs above.

Reply:
I've now removed these, to keep just `Projector{Diagonal}(...)`. If nothing else this saves a lot of pixels.